src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/DataContentAsyncEnumerableStream.cs

```diff
@@ -0,0 +1,228 @@
+    {
+        _enumerator = Throw.IfNull(dataAsyncEnumerable).GetAsyncEnumerator(cancellationToken);
+        _remainingData = Array.Empty<byte>();
+        _remainingDataOffset = 0;
+        _position = 0;
+        _firstDataContent = firstDataContent;
+    }
+
+    /// <inheritdoc/>
+    public override bool CanRead => true;
+
+    /// <inheritdoc/>
+    public override bool CanSeek => false;
+
+    /// <inheritdoc/>
+    public override bool CanWrite => false;
+
+    /// <inheritdoc/>
+    public override long Length => throw new NotSupportedException();
+
+    /// <inheritdoc/>
+    public override long Position
+    {
+        get => _position;
+        set => throw new NotSupportedException();
+    }
+
+    /// <inheritdoc/>
+    public override void Flush() => throw new NotSupportedException();
+
+    /// <inheritdoc/>
+    public override long Seek(long offset, SeekOrigin origin) =>
+        throw new NotSupportedException();
+
+    /// <inheritdoc/>
+    public override void SetLength(long value) =>
+        throw new NotSupportedException();
+
+    /// <inheritdoc/>
+    public override int Read(byte[] buffer, int offset, int count)
+    {
+        throw new NotSupportedException("Use ReadAsync instead for asynchronous reading.");
+    }
+
+    /// <inheritdoc/>
+#if NET8_0_OR_GREATER
+    public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        => ReadAsync(buffer, cancellationToken).AsTask();
+#else
+    public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+    {
+        if (_isCompleted)
+        {
+            return 0;
+        }
+
+        int bytesRead = 0;
+
+        while (bytesRead < count)
+        {
+            if (_remainingDataOffset < _remainingData.Length)
+            {
+                int bytesToCopy = Math.Min(count - bytesRead, _remainingData.Length - _remainingDataOffset);
+                Array.Copy(_remainingData, _remainingDataOffset, buffer, offset + bytesRead, bytesToCopy);
+                _remainingDataOffset += bytesToCopy;
+                bytesRead += bytesToCopy;
+                _position += bytesToCopy;
+            }
+            else
+            {
+                // Special case when the first chunk was skipped and needs to be read
+                if (_position == 0 && _firstDataContent is not null && _firstDataContent.Data.HasValue)
+                {
+                    _remainingData = _firstDataContent.Data.Value.ToArray();
+                    _remainingDataOffset = 0;
+
+                    continue;
+                }
+
+                if (!await _enumerator.MoveNextAsync().ConfigureAwait(false))
+                {
+                    _isCompleted = true;
+                    break;
+                }
+
+                if (!_enumerator.Current.Data.HasValue)
+                {
+                    _isCompleted = true;
```

Comment:
This isn't an error?

Reply:
In theory this should not happen, (the input send an empty chunk), and if that happens my first thought would, ok, if a get any chunk that is empty, I assume the stream just ended. Open for consideration, for stuff we should not expect to thru InvalidOperationException.